File: like-service/src/main/java/com/backend/like/service/PostLikeService.java
package com.backend.like.service;

import com.backend.like.dto.PostLikeDto;
import com.backend.like.model.PostLike;
import com.backend.like.repository.PostLikeRepository;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.web.bind.annotation.RequestBody;

import java.util.Optional;

@Service
@RequiredArgsConstructor
public class PostLikeService {

  private final PostLikeRepository postLikeRepository;

  @Transactional
  public String toggleLike(@RequestBody PostLikeDto postLikeDto) {
    Optional<PostLike> optionalPostLike = postLikeRepository.findByPostUUIDAndUserUUID(postLikeDto.postUUID(), postLikeDto.userUUID());

    String action;

    if (optionalPostLike.isPresent()) {

      PostLike existingLike = PostLike.builder()
          .postUUID(postLikeDto.postUUID())
          .userUUID(postLikeDto.userUUID())
          .build();
      postLikeRepository.delete(existingLike);
      action = "dislike";

    } else {
      PostLike postLike = PostLike.builder()
          .postUUID(postLikeDto.postUUID())
          .userUUID(postLikeDto.userUUID())
          .build();
      postLikeRepository.save(postLike);
      action = "like";
    }

    return action;

  }

}

File: like-service/src/main/kotlin/com/backend/like/service/PostLikeService.kt
package com.backend.like.service

import com.backend.like.dto.PostLikeDto
import com.backend.like.model.PostLike
import com.backend.like.repository.PostLikeRepository
import jakarta.transaction.Transactional
import jakarta.validation.Valid
import org.springframework.stereotype.Service
import org.springframework.web.bind.annotation.RequestBody

@Service
class PostLikeService(private val postLikeRepository: PostLikeRepository) {

    @Transactional
    fun toggleLike(@Valid @RequestBody postLikeDto: PostLikeDto): String {
        val existingLike: PostLike? =
            postLikeRepository.findByPostUUIDAndUserUUID(postLikeDto.postUUID, postLikeDto.userUUID)

        val action: String

        if (existingLike != null) {
            postLikeRepository.delete(existingLike)
            action = "dislike"
        } else {
            val like = PostLike(postUUID = postLikeDto.postUUID, userUUID = postLikeDto.userUUID)
            postLikeRepository.save(like)
            action = "like"
        }

        return action
    }
}
